ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/BrokerCompatibilityCheck.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright 2017 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package io.confluent.ksql.rest.server;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.UnsupportedVersionException;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.errors.StreamsException;
+
+import java.io.Closeable;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+import io.confluent.ksql.util.KsqlException;
+
+
+public class BrokerCompatibilityCheck implements Closeable {
+
+  private final Consumer<?, ?> consumer;
+  private final TopicPartition topicPartition;
+
+  BrokerCompatibilityCheck(final Consumer<?, ?> consumer,
+                           final TopicPartition topicPartition) {
+    this.consumer = consumer;
+    this.topicPartition = topicPartition;
+  }
+
+
+  public static BrokerCompatibilityCheck create(final Map<String, Object> streamsConfig, final Set<String> topicNames) {
+    // the offsetsForTime call needs a partition that exists else it can block forever
+    if (topicNames.isEmpty()) {
+      throw new KsqlException("Unable to check broker compatibility against a broker without any topics");
+    }
+    final Map<String, Object> consumerConfigs = new StreamsConfig(streamsConfig)
+        .getConsumerConfigs("__ksql_compatibility_check", "ksql_server");
+
+    // remove this otherwise it will try and instantiate the StreamsPartitionAssignor
+    consumerConfigs.remove("partition.assignment.strategy");
+    final KafkaConsumer<String, String> consumer
+        = new KafkaConsumer<>(consumerConfigs, new StringDeserializer(), new StringDeserializer());
+    return new BrokerCompatibilityCheck(consumer,  new TopicPartition(topicNames.iterator().next(), 0));
+  }
+
+  /**
+   * Check if the used brokers have version 0.10.1.x or higher.
+   *
+   * @throws KsqlException if brokers have version 0.10.0.x
+   */
+  void checkCompatibility() throws StreamsException {
+    try {
+      consumer.offsetsForTimes(Collections.singletonMap(topicPartition, 0L));
```

Comment:
If we have a mixed version cluster (like during an upgrade), this will validate the cluster even though some brokers are incompatible, right? Particularly, it only checks the version for the leader of partition 0.

We should file an issue to keep track of this limitation.. it may come back to bite us later.

Reply:
This would fix itself if it happens in the middle of an upgrade right? Not sure if a JIRA is needed, maybe just a comment in the code.

Reply:
It would fix itself for an upgrade. But not if the cluster was really mixed. 

Admittedly, the latter shouldn't happen, and actually being able to detect it reliably with this approach is probably impossible since you would have to know all the brokers, then create a topic with enough partitions, and then hope that the leaders are evenly distributed, etc.

So perhaps a comment denoting the limitation is enough.